tauri-updater/src/lib.rs

```diff
@@ -1,19 +1,1034 @@
+
+      // early finish we have everything we need here
+      return Ok(());
+    }
+  }
+
+  Ok(())
+}
+
+// Windows
+
+// ### Expected structure:
+// ├── [AppName]_[version]_x64.msi.zip          # ZIP generated by tauri-bundler
+// │   └──[AppName]_[version]_x64.msi           # Application MSI
+// └── ...
+
+// ## MSI
+// Update server can provide a MSI for Windows. (Generated with tauri-bundler from *Wix*)
+// To replace current version of the application. In later version we'll offer
+// incremental update to push specific binaries.
+
+// ## EXE
+// Update server can provide a custom EXE (installer) who can run any task.
+
+#[cfg(target_os = "windows")]
+fn copy_files_and_run(tmp_dir: tempfile::TempDir, _extract_path: PathBuf) -> Result {
+  let paths = read_dir(&tmp_dir).unwrap();
+  for path in paths {
+    let found_path = path.expect("Unable to extract").path();
+    // we support 2 type of files exe & msi for now
+    // If it's an `exe` we expect an installer not a runtime.
+    if found_path.extension() == Some(OsStr::new("exe")) {
+      // Run the EXE
+      tmp_dir.into_path();
+      Command::new(found_path)
+        .spawn()
+        .expect("installer failed to start");
+
+      exit(0);
+      // early finish we have everything we need here
+      return Ok(());
+    } else if found_path.extension() == Some(OsStr::new("msi")) {
+      // This consumes the TempDir without deleting directory on the filesystem,
+      // meaning that the directory will no longer be automatically deleted.
+      tmp_dir.into_path();
+
+      Command::new("msiexec.exe")
+        .arg("/i")
+        .arg(found_path)
+        .spawn()
+        .expect("installer failed to start");
+
+      exit(0);
+      // early finish we have everything we need here
+      return Ok(());
+    }
+  }
+
+  Ok(())
+}
+
+// MacOS
+
+// ### Expected structure:
+// ├── [AppName]_[version]_x64.app.tar.gz       # GZ generated by tauri-bundler
+// │   └──[AppName].app                         # Main application
+// │      └── Contents                          # Application contents...
+// │          └── ...
+// └── ...
+
+#[cfg(target_os = "macos")]
+fn copy_files_and_run(tmp_dir: tempfile::TempDir, extract_path: PathBuf) -> Result {
+  // In our tempdir we expect 1 directory (should be the <app>.app)
+  let paths = read_dir(&tmp_dir).unwrap();
+
+  for path in paths {
+    let found_path = path.expect("Unable to extract").path();
+    // make sure it's our .app
+    if found_path.extension() == Some(OsStr::new("app")) {
+      // Walk the temp dir and copy all files by replacing existing files only
+      // and creating directories if needed
+      Move::from_source(&found_path).walk_to_dest(&extract_path)?;
+
+      // early finish we have everything we need here
+      return Ok(());
+    }
+  }
+
+  Ok(())
+}
+
+/// Returns a target os -- If none,
+/// that mean the updater didnt support
+/// the platform
+pub fn get_target() -> Option<String> {
```

Comment:
Don't we have this logic somewhere else in the code base?  Maybe should import it from utils or something.